firewood/src/merkle/stream.rs

```diff
@@ -0,0 +1,505 @@
+    type Item = Result<(Vec<u8>, Vec<u8>), api::Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let MerkleKeyValueStream {
+            key_state,
+            merkle_root,
+            merkle,
+        } = &mut *self;
+
+        match key_state {
+            IteratorState::StartAtBeginning => {
+                let root = merkle
+                    .get_node(*merkle_root)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                // always put the sentinal node in parents
+                let mut parents = vec![(root, 0)];
+
+                let next_result = find_next_result(merkle, &mut parents, true)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                let (next_result, visit_last) = next_result
+                    .map(|(next, visit)| (Some(next), visit))
+                    .unwrap_or_default();
+
+                self.key_state = IteratorState::Iterating {
+                    visit_last,
+                    next_result,
+                    parents,
+                };
+
+                self.poll_next(_cx)
+            }
+
+            IteratorState::StartAtKey(key) => {
+                let root_node = merkle
+                    .get_node(*merkle_root)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                let (next_result, parents) = {
+                    let (found_node, parents) = {
+                        let mut parents = vec![];
+
+                        let found_node = merkle
+                            .get_node_by_key_with_callbacks(
+                                root_node,
+                                &key,
+                                |addr, nib| {
+                                    parents.push((addr, nib));
+                                },
+                                |_, _| {},
+                            )
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                        (found_node, parents)
+                    };
+
+                    let mut parents = parents
+                        .into_iter()
+                        .map(|(addr, nib)| merkle.get_node(addr).map(|node| (node, nib)))
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                    // means the keys definitely match
+                    if let Some(found_node) = found_node {
+                        let (value, visit_last) = match found_node.inner() {
+                            NodeType::Extension(_) => unreachable!(),
+
+                            NodeType::Branch(branch) => {
+                                let value = branch.value.as_ref().map(|v| v.to_vec());
```

Comment:
🤔, this is a little awkward. `get_node_by_key_with_callbacks` will only return nodes with data. 

This will go away once #399 is implemented.